### DISCUSSION TITLE
Guidance regarding problem converged to a point of local infeasibility

### DISCUSSION THREAD

**Discussion by tripiit16** [yalmip/YALMIP#1036] [Optimization modelling]
Dear Sir,
Please provide your suggestions regarding the local infeasibility of the non-linear problem solved using "ipopt" solver. 
Also, if I am changing the non-convex quadratic equality to SOCP and solving using ''gurobi", the problem becomes infeasible. Any suggestions will be highly appreciable. The objective I have kept constant for analysis.

**Comment by johanlofberg** [comment]
Why don't you simply try the nonconvex solver of gurobi

Otherwise, you will have to test some other local solver, such as fmincon

**Reply by johanlofberg** [reply]
and don't send code as Word files. No reason to not send the .m

**Reply by tripiit16** [reply]
.m file type is not supported, will send the zip file. If I convert the quadratic equality to SOCP constraints, instead of solving a non-convex problem. Will it improve the solution? Also, when I convert the quadratic equality to SOCP, it still shows GUROBI-NONCONVEX. Please
[New folder.zip](https://github.com/yalmip/YALMIP/files/7543253/New.folder.zip)
 find attached the zip file

**Reply by johanlofberg** [reply]
You cannot convert a quadratic equality to an SOCP, so your question and statement makes no sense.

If the solver is listed as GUROBI-NONCONVEX, it means YALMIP has detected a nonconvex quadratic program, and thus using the global nonconvex solver in gurobi.

**Reply by tripiit16** [reply]
Sorry, I was not able to convey it accurately. I am not converting quadratic equality to SOCP, I am relaxing it as an inequality constraint and using it as a cone.

**Reply by johanlofberg** [reply]
You still have nonconvex quadratics, otherwise it would not say GUROBI-NONCONVEX

**Reply by johanlofberg** [reply]
and the model is trivially detected as infeasible (it is infeasible before you start adding any quadratic stuff), and the nonconvexity is clear as you add stuff like

`co=[co,lij(:,2).^2<=lij(:,1).*lij(:,5)];`

 if that is intended to be a rotated SOCP cone, you are missing the convexifying sign restrictions like lij(:,1)>=0, lij(:,5)>=0

**Reply by tripiit16** [reply]
Thank you, sir, I have modified the constraints. If I don't relax these constraints and solve the model using "ipopt", the problem is converging to local infeasibility. However, after relaxing and solving using gurobi, the solution is infeasible. And yes it is detected infeasible way before adding quadratic constraints.

**Reply by johanlofberg** [reply]
Hence no reason to even talk or work with the quadratic stuff yet as it is apparently wrong already in the linear part

**Comment by tripiit16** [comment]
If a binary variable is in the model,  will Gurobi take the SOCP
constraints as quadratic constraints?

On Wed, Nov 17, 2021 at 7:45 PM Johan Löfberg ***@***.***>
wrote:

> Hence no reason to even talk or work with the quadratic stuff yet as it is
> apparently wrong already in the linear part
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yalmip/YALMIP/discussions/1036#discussioncomment-1658318>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQOJ5RLQAPVNLL5G2AIYL6TUMO2IBANCNFSM5IBWH5IA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Regards
Tripti Gangwar
Research Scholar
Electrical Engineering Department
IIT Roorkee

**Reply by johanlofberg** [reply]
Huh? An SOCP is an SOCP (sometimes sloppily called a quadratic constraint although that is something less general) regardless of the existence of binaries or not

**Comment by tripiit16** [comment]
The attached problem shows 117 quadratic constraints if there is a binary
variable in the problem. However, if I relax or remove that binary
variable, the log shows 117second order cone constraints.  I know quadratic
is generalized SOCP, then why is the log different? I mean, Out of
curiosity, I thought so

On Fri, Nov 26, 2021 at 12:55 PM Johan Löfberg ***@***.***>
wrote:

> Huh? An SOCP is an SOCP (sometimes sloppily called a quadratic constraint
> although that is something less general) regardless of the existence of
> binaries or not
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yalmip/YALMIP/discussions/1036#discussioncomment-1703946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQOJ5RNG2R4RBOGIPP6NPXTUN4Y7XANCNFSM5IBWH5IA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Regards
Tripti Gangwar
Research Scholar
Electrical Engineering Department
IIT Roorkee

clc
clear all
Vbase = 4.160/sqrt(3);
Sbase = 5;
Linedata=xlsread('IEEE13Nodeckt_LOADSn.xlsx','sheet3');
% Linedata=[Linedata(:,(1:2)) Linedata(:,(3:20))/Zbase];
% Linedata(5,3:20)=Linedata(5,3:20)*Zbase;
r=Linedata(:,(3:11));
x=Linedata(:,(12:20));


Load = [
1      0                0              0;      %650 grid
2      0		        0		       0		; ... % 632
3    (385+220i)/3	(385+220i)/3	(385+220i)/3	; ... % 671   Delta
4      0                            0                  0;
5      0                            0                  0;
6   160+110i		120+90i		120+90i		; ... %  634
7        0			170+125i		0			; ... %645
8        0			(230+132i)	 0	; ... % 646       Delta
9         0 	0			(170+151i)	; ... % 692   Delta
10      485+190i		68+60i		290+212i		; ... %  675
11       0                            0                  0;
12         0			0			170+80i		; ... %611
13     128+86i		0			0			;    %652
14      17+10i      66+38i    117+68i];    %670
Load= Load(:,(2:4))/(Sbase*333.33);

%Capacitor bank
Cap= [
1      0                0              0;      %650 grid
2      0		        0		       0		; ... % 632
3      0		        0		       0	; ... % 671  
4      0                            0                  0;
5      0                            0                  0;
6      0		        0		       0	; ... %  634
7        0			0		0			; ... %645
8        0		        0		       0	; ... % 646     
9        0		        0		       0; ... % 692   
10     200	    200          	200		; ... %  675
11       0                            0                  0;
12         0			0			100		; ... %611
13      0		0			0			;    %652
14     0           0           0];    %670

Cap=Cap(:,(2:4))/(Sbase*333.33);

%%%current angles from opendss

ang_m=xlsread('IEEE13Nodeckt_LOADSn.xlsx','sheet4');
ang_m=ang_m(:,3:11);
M=100;
Vmin=0.95;
Vmax=1.05;
a = exp(1j*2*pi/3);

%%%% Variables
nb=max(Linedata(:,2));   % number of buses
nbr=length(Linedata(:,1));  % number of branches
ph=3; %set of phases
phij=9; %set of pp,pq
pgrid=sdpvar(1,ph,'full');
qgrid=sdpvar(1,ph,'full');
v=sdpvar(nb,ph,'full');  %square of voltage
lij=sdpvar(nbr,phij,'full'); %square of current
% pij=sdpvar(nbr,phij,'full');  %active branch flow
% qij=sdpvar(nbr,phij,'full');   %reactive branch flow
pij=sdpvar(nbr,ph,'full');  %active branch flow
qij=sdpvar(nbr,ph,'full');   %reactive branch flow
pinj=sdpvar(nb,ph,'full');      %active power injection at each bus
qinj=sdpvar(nb,ph,'full');      %reactive power injection at each bus
qcap=sdpvar(nb,ph,'full');      %reactive power generated by capacitor bank
 ucap=binvar(nb,1,'full');     %status of cap bank
%   ucap=sdpvar(nb,ph,'full');  
zcap=sdpvar(nb,1,'full');  %linearization variable
%%%% network constraints
co=[];

%%%%power flow equations
  aux=zeros(nb,nbr);
 
    for i=1:nbr
        aux(Linedata(i,1),i)=1;
        aux(Linedata(i,2),i)=-1;
    end
    ucap=[ucap ucap ucap];
    zcap=[zcap zcap zcap];
 pinj=-real(Load);
    qinj=(zcap.*Cap)-imag(Load);
    
    co=[co,zcap<=M*ucap];
co=[co,zcap<=v];
co=[co,zcap>=v-(1-ucap)*M];
co=[co,zcap>=0];

   for i=1:3
  kp(:,i)=(r(:,i).*cos(ang_m(:,i))+x(:,i).*sin(ang_m(:,i))).*lij(:,i);
  kp1(:,i)=(r(:,i+3).*cos(ang_m(:,i+3))+x(:,i+3).*sin(ang_m(:,i+3))).*lij(:,i+3);
  kp2(:,i)=(r(:,i+6).*cos(ang_m(:,i+6))+x(:,i+6).*sin(ang_m(:,i+6))).*lij(:,i+6);
  kq(:,i)=(x(:,i).*cos(ang_m(:,i))-r(:,i).*sin(ang_m(:,i))).*lij(:,i);
  kq1(:,i)=(x(:,i+3).*cos(ang_m(:,i+3))-r(:,i+3).*sin(ang_m(:,i+3))).*lij(:,i+3);
  kq2(:,i)=(x(:,i+6).*cos(ang_m(:,i+6))-r(:,i+6).*sin(ang_m(:,i+6))).*lij(:,i+6);
   end 
     
co=[co,pij(1,:)==pgrid]; 
co=[co,qij(1,:)==qgrid]; 

  co=[co,(aux(2:nb,:)*pij(:,1))-pinj(2:nb,1)+sum(kp,2)==0];
    co=[co,(aux(2:nb,:)*pij(:,2))-pinj(2:nb,2)+sum(kp1,2)==0];
   co=[co,(aux(2:nb,:)*pij(:,3))-pinj(2:nb,3)+sum(kp2,2)==0];
  co=[co,(aux(2:nb,:)*qij(:,1))-qinj(2:nb,1)+sum(kq,2)==0];
  co=[co,(aux(2:nb,:)*qij(:,2))-qinj(2:nb,2)+sum(kq1,2)==0];
   co=[co,(aux(2:nb,:)*qij(:,3))-qinj(2:nb,3)+sum(kq2,2)==0];   
     
% %%%%voltage equation

k=2*(r(:,1).*pij(:,1)+x(:,1).*qij(:,1))+(pij(:,2).*(sqrt(3)*x(:,2)-r(:,2)))+(pij(:,3).*(-sqrt(3)*x(:,3)-r(:,3)))+(qij(:,2).*(-x(:,2)-sqrt(3)*r(:,2)))+(qij(:,3).*(sqrt(3)*r(:,3)-x(:,3)));
k1=2*(r(:,5).*pij(:,2)+x(:,5).*qij(:,2))+(pij(:,1).*(-sqrt(3)*x(:,4)-r(:,4)))+(pij(:,3).*(sqrt(3)*x(:,6)-r(:,6)))+(qij(:,1).*(-x(:,4)+sqrt(3)*r(:,4)))+(qij(:,3).*(-sqrt(3)*r(:,6)-x(:,6)));
k2=2*(r(:,9).*pij(:,3)+x(:,9).*qij(:,3))+(pij(:,2).*(-sqrt(3)*x(:,8)-r(:,8)))+(pij(:,1).*(sqrt(3)*x(:,7)-r(:,7)))+(qij(:,2).*(-x(:,8)+sqrt(3)*r(:,8)))+(qij(:,1).*(-sqrt(3)*r(:,7)-x(:,7)));

ka=(2*lij(:,3).*((r(:,1).*r(:,3)+x(:,1).*x(:,3)).*cos(ang_m(:,3))+(r(:,1).*x(:,3)-r(:,3).*x(:,1)).*sin(ang_m(:,3))))+(2*lij(:,2).*((r(:,1).*r(:,2)+x(:,1).*x(:,2)).*cos(ang_m(:,2))+(r(:,1).*x(:,2)-r(:,2).*x(:,1)).*sin(ang_m(:,2))))+(2*lij(:,6).*((r(:,3).*r(:,2)+x(:,3).*x(:,2)).*cos(ang_m(:,6))+(r(:,2).*x(:,3)-r(:,3).*x(:,2)).*sin(ang_m(:,6))));
kb=(2*lij(:,3).*((r(:,4).*r(:,6)+x(:,4).*x(:,6)).*cos(ang_m(:,3))+(r(:,4).*x(:,6)-r(:,6).*x(:,4)).*sin(ang_m(:,3))))+(2*lij(:,2).*((r(:,4).*r(:,5)+x(:,4).*x(:,5)).*cos(ang_m(:,2))+(r(:,4).*x(:,5)-r(:,5).*x(:,4)).*sin(ang_m(:,2))))+(2*lij(:,6).*((r(:,5).*r(:,6)+x(:,5).*x(:,6)).*cos(ang_m(:,6))+(r(:,5).*x(:,6)-r(:,6).*x(:,5)).*sin(ang_m(:,6))));
kc=(2*lij(:,3).*((r(:,7).*r(:,9)+x(:,7).*x(:,9)).*cos(ang_m(:,3))+(r(:,7).*x(:,9)-r(:,9).*x(:,7)).*sin(ang_m(:,3))))+(2*lij(:,2).*((r(:,7).*r(:,8)+x(:,7).*x(:,8)).*cos(ang_m(:,2))+(r(:,7).*x(:,8)-r(:,8).*x(:,7)).*sin(ang_m(:,2))))+(2*lij(:,6).*((r(:,8).*r(:,9)+x(:,8).*x(:,9)).*cos(ang_m(:,6))+(r(:,8).*x(:,9)-r(:,9).*x(:,8)).*sin(ang_m(:,6))));
  
  co=[co,v(Linedata(:,2),1)==v(Linedata(:,1),1)-k+(r(:,1).^2+x(:,1).^2).*lij(:,1)+(r(:,2).^2+x(:,2).^2).*lij(:,5)+(r(:,3).^2+x(:,3).^2).*lij(:,9)+ka];
  co=[co,v(Linedata(:,2),2)==v(Linedata(:,1),2)-k1+(r(:,4).^2+x(:,4).^2).*lij(:,1)+(r(:,5).^2+x(:,5).^2).*lij(:,5)+(r(:,6).^2+x(:,6).^2).*lij(:,9)+kb];
  co=[co,v(Linedata(:,2),3)==v(Linedata(:,1),3)-k2+(r(:,7).^2+x(:,7).^2).*lij(:,1)+(r(:,8).^2+x(:,8).^2).*lij(:,5)+(r(:,9).^2+x(:,9).^2).*lij(:,9)+kc];


%    co=[co,pij(:,2)==-0.5*(pij(:,5))-0.866*(qij(:,5))];
%       co=[co,pij(:,3)==-0.5*(pij(:,9))+0.866*(qij(:,9))];
%        co=[co,pij(:,4)==-0.5*(pij(:,1))+0.866*(qij(:,1))];
%         co=[co,pij(:,6)==-0.5*(pij(:,9))-0.866*(qij(:,9))];
%        co=[co,pij(:,8)==-0.5*(pij(:,5))+0.866*(qij(:,5))];
%       co=[co,pij(:,7)==-0.5*(pij(:,1))-0.866*(qij(:,1))];
%       co=[co,qij(:,2)==-0.5*(qij(:,5))+0.866*(pij(:,5))];
%       co=[co,qij(:,3)==-0.5*(qij(:,9))-0.866*(pij(:,9))];
%        co=[co,qij(:,4)==-0.5*(qij(:,1))--0.866*(pij(:,1))];
%         co=[co,qij(:,6)==-0.5*(qij(:,9))+0.866*(pij(:,9))];
%        co=[co,qij(:,8)==-0.5*(qij(:,5))-0.866*(pij(:,5))];
%        co=[co,qij(:,7)==-0.5*(qij(:,1))+0.866*(pij(:,1))];
      
%         co=[co,lij(:,2).^2==lij(:,1).*lij(:,5)];
%       co=[co,lij(:,3).^2==lij(:,1).*lij(:,9)];
%      co=[co,lij(:,4).^2==lij(:,1).*lij(:,5)];
%       co=[co,lij(:,6).^2==lij(:,5).*lij(:,9)];  
%        co=[co,lij(:,7).^2==lij(:,1).*lij(:,9)];
%       co=[co,lij(:,8).^2==lij(:,5).*lij(:,9)];
          co=[co,lij(:,2).^2<=lij(:,1).*lij(:,5)];
      co=[co,lij(:,3).^2<=lij(:,1).*lij(:,9)];
     co=[co,lij(:,4).^2<=lij(:,1).*lij(:,5)];
      co=[co,lij(:,6).^2<=lij(:,5).*lij(:,9)];  
       co=[co,lij(:,7).^2<=lij(:,1).*lij(:,9)];
      co=[co,lij(:,8).^2<=lij(:,5).*lij(:,9)];
 co=[co,lij>=0];
 co=[co,v>=0];
      
    co=[co,-10/Sbase<=pgrid<=10/Sbase];
    co=[co,-10/Sbase<=qgrid<=10/Sbase];
      co = [co, (Vmin)^2<= v<=(Vmax)^2];  
      co=[co,v(1,:)==1];
          co=[co,-4/Sbase<=pij<=4/Sbase];
      co=[co,-4/Sbase<=qij<=4/Sbase];
      
     co=[ co,pij(:,1).^2+qij(:,1).^2<=v(2:nb,1).*lij(:,1)];
     co=[ co,pij(:,2).^2+qij(:,2).^2<=v(2:nb,2).*lij(:,5)];
     co=[ co,pij(:,3).^2+qij(:,3).^2<=v(2:nb,3).*lij(:,9)];
% 
%      co=[ co,pij(:,1).^2+qij(:,1).^2==v(2:nb,1).*lij(:,1)];
%      co=[ co,pij(:,2).^2+qij(:,2).^2==v(2:nb,2).*lij(:,5)];
%      co=[ co,pij(:,3).^2+qij(:,3).^2==v(2:nb,3).*lij(:,9)];

     %%%Objective
Objective=sum(pgrid);
ops = sdpsettings('verbose',1,'debug',1,'solver','gurobi');
sol= optimize(co,Objective,ops)

**Reply by johanlofberg** [reply]
My guess is Gurobi is inconsistent in naming. Here it writes quadratic for an socp model

>> optimize(norm(x) <= 1,-sum(x))
Academic license - for non-commercial use only - expires 2022-09-30
Gurobi Optimizer version 9.1.2 build v9.1.2rc0 (win64)
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 5 rows, 8 columns and 9 nonzeros
Model fingerprint: 0x46eeb0ed
Model has 1 quadratic constraint

**Reply by tripiit16** [reply]
Presolve shows one second order cone constraint

Model has 1 quadratic constraint
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  QMatrix range    [1e+00, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+00]
Presolve removed 3 rows and 3 columns
Presolve time: 0.04s
Presolved: 1 rows, 2 columns, 1 nonzeros
**Presolved model has 1 second-order cone constraint**
Ordering time: 0.00s
